File: solutions/fortran/saddle-points/1/saddle_points.f90
module saddle_points
  implicit none

  type :: point_t
    integer :: row
    integer :: column
  end type point_t

contains

  function saddlePoints(matrix) result(points)
    integer, intent(in) :: matrix(:, :)
    type(point_t), allocatable :: points(:)

    integer :: nrows, ncols, i, j, cnt, row_max
    integer, allocatable :: col_min(:)
    type(point_t), allocatable :: temp(:)

    nrows = size(matrix, 1)
    ncols = size(matrix, 2)

    if (nrows == 0 .or. ncols == 0) then
      allocate(points(0))
      return
    end if

    ! Precompute column minima
    allocate(col_min(ncols))
    do j = 1, ncols
      col_min(j) = matrix(1, j)
      do i = 2, nrows
        if (matrix(i, j) < col_min(j)) col_min(j) = matrix(i, j)
      end do
    end do

    ! Temporary storage (maximum possible number of points)
    allocate(temp(nrows * ncols))
    cnt = 0

    do i = 1, nrows
      ! Find row maximum
      row_max = matrix(i, 1)
      do j = 2, ncols
        if (matrix(i, j) > row_max) row_max = matrix(i, j)
      end do

      ! Identify saddle points in this row
      do j = 1, ncols
        if (matrix(i, j) == row_max .and. matrix(i, j) == col_min(j)) then
          cnt = cnt + 1
          temp(cnt) = point_t(row = i, column = j)
        end if
      end do
    end do

    ! Trim to the actual number of points
    allocate(points(cnt))
    if (cnt > 0) then
      points(1:cnt) = temp(1:cnt)
    end if
  end function saddlePoints

end module saddle_points

File: solutions/c/saddle-points/1/saddle_points.c
#include "saddle_points.h"
#include <stdlib.h>

saddle_points_t *saddle_points(int rows, int cols, uint8_t matrix[rows][cols]) {
    saddle_points_t *result = malloc(sizeof(saddle_points_t));
    if (!result) return NULL;

    size_t max_points = (size_t)rows * cols;
    result->points = malloc(sizeof(saddle_point_t) * max_points);
    if (!result->points && max_points > 0) {
        free(result);
        return NULL;
    }
    result->count = 0;

    if (rows <= 0 || cols <= 0) {
        return result; // empty matrix
    }

    // Row maximums
    uint8_t row_max[rows];
    for (int i = 0; i < rows; i++) {
        uint8_t max_val = matrix[i][0];
        for (int j = 1; j < cols; j++) {
            if (matrix[i][j] > max_val) {
                max_val = matrix[i][j];
            }
        }
        row_max[i] = max_val;
    }

    // Column minimums
    uint8_t col_min[cols];
    for (int j = 0; j < cols; j++) {
        uint8_t min_val = matrix[0][j];
        for (int i = 1; i < rows; i++) {
            if (matrix[i][j] < min_val) {
                min_val = matrix[i][j];
            }
        }
        col_min[j] = min_val;
    }

    // Collect saddle points (1‑based indices)
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (matrix[i][j] == row_max[i] && matrix[i][j] == col_min[j]) {
                result->points[result->count].row = (size_t)(i + 1);
                result->points[result->count].column = (size_t)(j + 1);
                result->count++;
            }
        }
    }

    // Trim to exact size
    if (result->count > 0) {
        saddle_point_t *tmp = realloc(result->points, sizeof(saddle_point_t) * result->count);
        if (tmp) {
            result->points = tmp;
        }
    } else {
        free(result->points);
        result->points = NULL;
    }

    return result;
}

void free_saddle_points(saddle_points_t *sp) {
    if (sp) {
        free(sp->points);
        free(sp);
    }
}